homeassistant/components/notify/matrix.py

```diff
@@ -0,0 +1,177 @@
+                if not self.login_by_token():
+                    if not self.login_by_password():
+                        _LOGGER.error(
+                            'login failed, both token and username/password '
+                            'invalid'
+                        )
+                        return
+            else:
+                if not self.login_by_password():
+                    _LOGGER.error('login failed, username/password invalid')
+                    return
+
+            rooms = self.client.get_rooms()
+            for target_room in target_rooms:
+                try:
+                    if target_room in rooms:
+                        room = rooms[target_room]
+                    else:
+                        room = self.client.join_room(target_room)
+
+                    _LOGGER.debug(room.send_text(message))
+                except MatrixRequestError as ex:
+                    _LOGGER.error(
+                        'Unable to deliver message to room \'%s\': (%d): %s',
+                        target_room, ex.code, ex.content
+                    )
+
+        def load_token(self):
+            """Load authentication token from persistent storage, if exists."""
+            if not os.path.exists(self.session_file):
+                return
+
+            with open(self.session_file) as handle:
+                data = json.load(handle)
+
+            try:
+                self.token = data['token']
+            except KeyError:
+                pass
+
+        def store_token(self, token):
+            """Store authentication token to persistent storage."""
+            with open(self.session_file, 'w') as handle:
+                handle.write(json.dumps({
+                    'token': token
+                }))
+
+        def login_by_token(self):
+            """Login using authentication token."""
+            try:
+                self.client = MatrixClient(
+                    homeserver,
+                    token=self.token,
+                    valid_cert_check=verify_tls
+                )
+                return True
+            except MatrixRequestError as ex:
+                _LOGGER.info(
+                    'login_by_token: (%d) %s', ex.code, ex.content
+                )
+                return False
+
+        def login_by_password(self):
+            """Login using password authentication."""
+            try:
+                self.client = MatrixClient(
+                    homeserver,
+                    valid_cert_check=verify_tls
+                )
+                self.store_token(
+                    self.client.login_with_password(
+                        username, password
+                    )
+                )
+                return True
+            except MatrixRequestError as ex:
+                _LOGGER.error(
+                    'login_by_password: (%d) %s', ex.code, ex.content
+                )
+                return False
+
+    MatrixNotificationClient()
```

Comment:
Not sure I understand why you use a class to send the message. Cant you move the contents of the init function here and make load_toke/store_token normal functions with return values?